src/listview.tsx

```diff
@@ -1,5 +1,54 @@
 import * as O from "./observable/observables"
-import { DOMNode, HarmajaOutput, HarmajaStaticOutput, LowLevelApi as H } from "./harmaja"
+import { DOMNode, HarmajaOutput, HarmajaStaticOutput, LowLevelApi as H, debug } from "./harmaja"
+
+// Find starting from hint
+function findIndex<A>(xs: A[], test: (value: A) => boolean, hint: number): number {
+    const len = xs.length
+    let u = hint
+    if (u >= len) u = len - 1
+    if (u < 0) u = 0
+    let d = u - 1
+    for (; 0 <= d && u < len; ++u, --d) {
+        if (test(xs[u])) return u
+        if (test(xs[d])) return d
+    }
+    for (; u < len; ++u) {
+        if (test(xs[u])) return u
+    }
+    for (; 0 <= d; --d) {
+        if (test(xs[d])) return d
+    }
+
+    // Not found
+    return -1
+}
+
+function findKey<A, K>(getKey: (value: A) => K, expected: K): O.Lens<A[], A | undefined> {
+    const test = (x: A) => getKey(x) === expected
+
+    // Cache the previous index. When items are moved they tend to end up near
+    // the previous index, it makes the search faster to start from the previous
+    // index and move down and up from there.
+    let index: number = -1
+
+    return {
+        get(root) {
+            index = findIndex(root, test, index)
+            return index >= 0 ? root[index] : undefined
+        },
+        set(root, value) {
+            index = findIndex(root, test, index)
+            if (index === -1) return root
+            if (value === undefined) {
+                // This peculiar lens deletes the item when undefined is written
+                const result = root.slice(0)
+                result.splice(index, 1)
+                return result
+            }
+            return [...root.slice(0, index), value, ...root.slice(index + 1)]
+        }
+    }
+}
```

Comment:
The idea of caching the previous index and using that as a starting point was inspired by [`L.find` of partial.lenses](https://github.com/calmm-js/partial.lenses#l-find)